<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="CSHstats"><title>distances, nearest-neighbor graphs, clustering • CSHstats</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Roboto-0.4.1/font.css" rel="stylesheet"><link href="../deps/_Courier%20Prime-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="distances, nearest-neighbor graphs, clustering"><meta property="og:description" content="CSHstats"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CSHstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.13</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/S1_probability.html">S1_probability: Definitions and exercises</a>
    <a class="dropdown-item" href="../articles/S2_EDA.html">S2_EDA: Exploratory data analysis and visualization</a>
    <a class="dropdown-item" href="../articles/S3_distclust.html">S3: Distances, clustering</a>
    <a class="dropdown-item" href="../articles/S4_hypothesis_testing.html">S3_inference: Definitions and exercises</a>
    <a class="dropdown-item" href="../articles/S5_linear_models.html">S4_inference: Linear models and GLMs</a>
    <a class="dropdown-item" href="../articles/distplus.html">distances, nearest-neighbor graphs, clustering</a>
    <a class="dropdown-item" href="../articles/scCluster.html">UNKNOWN TITLE</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/vjcitn/CSHstats/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>distances, nearest-neighbor graphs, clustering</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">July 02, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/CSHstats/blob/HEAD/vignettes/distplus.Rmd" class="external-link"><code>vignettes/distplus.Rmd</code></a></small>
      <div class="d-none name"><code>distplus.Rmd</code></div>
    </div>

    
    
<div id="road-map" class="section level1">
<h1>Road map</h1>
<ul>
<li>Concepts underlying cluster analysis
<ul>
<li>Distances in high-dimensional spaces</li>
<li>Criteria for agglomerative clustering</li>
</ul></li>
<li>Exploring a cluster analysis of copy-number aberration measures in 28 tumors from one patient
<ul>
<li>interactive heatmap</li>
<li>comparing Euclidean and correlation (1-cor) distances</li>
<li>comparing agglomeration methods</li>
<li>the silhouette as a measure of clustering adequacy</li>
<li>the bluster tools for cluster diagnostics</li>
</ul></li>
<li>bluster applied to scRNA-seq
<ul>
<li>Grun pancreas data</li>
<li>NNGraphParam</li>
</ul></li>
</ul>
</div>
<div id="cluster-analysis-concepts" class="section level1">
<h1>Cluster analysis concepts</h1>
<p>The need to identify groups of similar observations arises in many contexts – ultimately in the service of clarifying sources of variability, building power for statistical comparisons.</p>
<p>Hierarchical clustering of N multivariate observations is conducted by starting with N clusters and proceeding from M clusters to M-1 (M=N, …, 2) clusters by merging the two members separated by the least of all pairwise distances.</p>
<p><img src="/home/stvjc/CSHstats/vignettes/agglomPic.jpg" /></p>
<div id="distances-in-high-dimensional-space" class="section level2">
<h2>Distances in high-dimensional space</h2>
<p><img src="/home/stvjc/CSHstats/vignettes/distances.jpg" /></p>
</div>
<div id="criteria-for-agglomerative-clustering" class="section level2">
<h2>Criteria for agglomerative clustering</h2>
<p><img src="/home/stvjc/CSHstats/vignettes/agglom.jpg" /></p>
</div>
</div>
<div id="application-inferring-steps-in-tumor-metastasis-in-a-breast-cancer-patient" class="section level1">
<h1>Application: inferring steps in tumor metastasis in a breast cancer patient</h1>
<p>We’ll examine data distributed with a <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-021-00989-6">2021 Genome Biology paper</a> from the Gabor Marth lab.</p>
<div class="figure">
<img src="/home/stvjc/CSHstats/vignettes/marthPatient.jpg" alt="Clinical sequence of interventions." />
<p class="caption">Clinical sequence of interventions.</p>
</div>
<div class="figure">
<img src="/home/stvjc/CSHstats/vignettes/marthEvolve.jpg" alt="Event sequence." />
<p class="caption">Event sequence.</p>
</div>
<div id="a-view-of-copy-number-aberrations-for-1mb-tiling" class="section level2">
<h2>A view of copy number aberrations for 1Mb tiling</h2>
<p>28 tumors were sampled and sequenced in a rapid autopsy procedure. Copy number variation was assessed using <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5027494/">FACETS</a>.</p>
<p>The tissues from which tumors were taken are Br (Breast), Bo (Bone), Bn (Brain), Ln (Lung), Lv (Liver), Pa (Pancreas), Ly (Lymph nodes), Kd (Kidney)</p>
<p><a href="cnv-heatmap.html">A plotly-based visualization</a></p>
<p>The (vertical) ordering of tissues is chosen to exemplify certain similarities.</p>
<p>For example the block of blue on chr10 is seen for only three samples. This is an indication of a deletion.</p>
</div>
<div id="a-cluster-analysis-proposed-in-support-of-the-evolutionary-map" class="section level2">
<h2>A cluster analysis proposed in support of the evolutionary map</h2>
<p>This code is lightly modified from a script distributed at <a href="https://github.com/xiaomengh/tumor-evo-rapid-autopsy.git" class="uri">https://github.com/xiaomengh/tumor-evo-rapid-autopsy.git</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">suppressPackageStartupMessages</span>({</a>
<a class="sourceLine" id="cb1-2" title="2"> <span class="kw">library</span>(csamaDist)</a>
<a class="sourceLine" id="cb1-3" title="3"> <span class="kw">library</span>(bioDist)</a>
<a class="sourceLine" id="cb1-4" title="4"> <span class="kw">library</span>(bluster)</a>
<a class="sourceLine" id="cb1-5" title="5"> <span class="kw">library</span>(cluster)</a>
<a class="sourceLine" id="cb1-6" title="6"> <span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb1-7" title="7"> <span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb1-8" title="8"> <span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb1-9" title="9"> <span class="kw">library</span>(scuttle)</a>
<a class="sourceLine" id="cb1-10" title="10">})</a>
<a class="sourceLine" id="cb1-11" title="11"><span class="kw">data</span>(cnv_log_R)</a>
<a class="sourceLine" id="cb1-12" title="12">data =<span class="st"> </span>cnv_log_R</a>
<a class="sourceLine" id="cb1-13" title="13">samples =<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;Ln7&#39;</span>,<span class="st">&#39;Ln9&#39;</span>,<span class="st">&#39;Ln1&#39;</span>,<span class="st">&#39;BrM&#39;</span>,<span class="st">&#39;BrP&#39;</span>,</a>
<a class="sourceLine" id="cb1-14" title="14">           <span class="st">&#39;Ln11&#39;</span>,<span class="st">&#39;Ly2&#39;</span>,<span class="st">&#39;Ln3&#39;</span>,</a>
<a class="sourceLine" id="cb1-15" title="15">           <span class="st">&#39;Bo3&#39;</span>,<span class="st">&#39;Ln10&#39;</span>,<span class="st">&#39;Bo1&#39;</span>,<span class="st">&#39;Ln8&#39;</span>,<span class="st">&#39;Lv3&#39;</span>,<span class="st">&#39;Ln5&#39;</span>,<span class="st">&#39;Bo2&#39;</span>,<span class="st">&#39;Bn2&#39;</span>,<span class="st">&#39;Bn1&#39;</span>,<span class="st">&#39;Bn3&#39;</span>,<span class="st">&#39;Bn4&#39;</span>,<span class="st">&#39;Ln2&#39;</span>,</a>
<a class="sourceLine" id="cb1-16" title="16">           <span class="st">&#39;Ly1&#39;</span>,<span class="st">&#39;Ln6&#39;</span>,</a>
<a class="sourceLine" id="cb1-17" title="17">           <span class="st">&#39;Kd1&#39;</span>,<span class="st">&#39;Ln4&#39;</span>,<span class="st">&#39;Lv4&#39;</span>,<span class="st">&#39;Lv2&#39;</span>,<span class="st">&#39;Lv1&#39;</span>,<span class="st">&#39;Pa1&#39;</span>)</a>
<a class="sourceLine" id="cb1-18" title="18"><span class="kw">rownames</span>(data) =<span class="st"> </span>samples</a>
<a class="sourceLine" id="cb1-19" title="19">d =<span class="st"> </span><span class="kw">dist</span>(data, <span class="dt">method=</span><span class="st">&quot;euclidean&quot;</span>)</a>
<a class="sourceLine" id="cb1-20" title="20">fit =<span class="st"> </span><span class="kw">hclust</span>(d, <span class="dt">method=</span><span class="st">&quot;average&quot;</span>)</a>
<a class="sourceLine" id="cb1-21" title="21"><span class="co"># the following line changes the order of the samples to produce the Fig.S3B but doesn&#39;t change the phylogenetic relationship</span></a>
<a class="sourceLine" id="cb1-22" title="22">fit<span class="op">$</span>order =<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">3</span>,<span class="dv">13</span>,<span class="dv">10</span>,<span class="dv">20</span>,<span class="dv">16</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">15</span>,<span class="dv">9</span>,<span class="dv">17</span>,<span class="dv">19</span>,<span class="dv">14</span>,<span class="dv">18</span>,<span class="dv">21</span>,<span class="dv">22</span>,<span class="dv">7</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">25</span>,<span class="dv">27</span>,<span class="dv">26</span>,<span class="dv">28</span>,<span class="dv">23</span>,<span class="dv">24</span>)</a>
<a class="sourceLine" id="cb1-23" title="23"><span class="kw">plot</span>(fit)</a></code></pre></div>
<p><img src="distplus_files/figure-html/doclus-1.png" width="700" /></p>
</div>
<div id="drilling-down-on-the-clustering" class="section level2">
<h2>Drilling down on the clustering</h2>
<div id="comparing-euclidean-and-correlation-distances" class="section level3">
<h3>Comparing Euclidean and Correlation distances</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">cd =<span class="st"> </span><span class="kw">cor.dist</span>(cnv_log_R) <span class="co"># from bioDist</span></a>
<a class="sourceLine" id="cb2-2" title="2">ed =<span class="st"> </span><span class="kw">dist</span>(cnv_log_R)</a>
<a class="sourceLine" id="cb2-3" title="3"><span class="kw">plot</span>(<span class="kw">as.numeric</span>(ed), <span class="kw">as.numeric</span>(cd), <span class="dt">xlab=</span><span class="st">&quot;All pairwise Euclidean distances&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;All pairwise correlation distances&quot;</span>)</a></code></pre></div>
<p><img src="distplus_files/figure-html/dodists-1.png" width="700" /></p>
<p>For a given correlation distance value, there can be wide variation in euclidean distance, and vice versa.</p>
<p>Open question: What distance metric is most relevant for biological interpretation of CNV?</p>
</div>
<div id="a-pair-with-discrepant-correlation-and-euclidean-distance-values-over-entire-genome" class="section level3">
<h3>A pair with discrepant correlation and euclidean distance values (over entire genome)</h3>
<p>We’ll have a look at the first 100Mb on chr1.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw">plot</span>(cnv_log_R[<span class="st">&quot;Ly1&quot;</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>],<span class="dt">pch=</span><span class="dv">19</span>, <span class="dt">main=</span><span class="st">&quot;chr1, first 100Mb&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;FACETS CNV log R&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;chr1&quot;</span>)</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">points</span>(cnv_log_R[<span class="st">&quot;Ln1&quot;</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>], <span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">pch=</span><span class="dv">19</span>)</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="kw">legend</span>(<span class="dv">60</span>, <span class="fl">-.5</span>, <span class="dt">pch=</span><span class="dv">19</span>, <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Ly1&quot;</span>, <span class="st">&quot;Ln1&quot;</span>))</a></code></pre></div>
<p><img src="distplus_files/figure-html/lk1mb-1.png" width="700" /></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co">#cor(cnv_log_R[&quot;Ly1&quot;, 1:100], cnv_log_R[&quot;Ln1&quot;, 1:100])</span></a>
<a class="sourceLine" id="cb4-2" title="2">edist =<span class="st"> </span><span class="cf">function</span>(x,y) <span class="kw">sqrt</span>(<span class="kw">sum</span>((x<span class="op">-</span>y)<span class="op">^</span><span class="dv">2</span>))</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw">edist</span>(cnv_log_R[<span class="st">&quot;Ly1&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>], cnv_log_R[<span class="st">&quot;Ln1&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>])</a></code></pre></div>
<pre><code>## [1] 1.685992</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw">plot</span>(<span class="kw">jitter</span>(cnv_log_R[<span class="st">&quot;Ly1&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>]), cnv_log_R[<span class="st">&quot;Ln1&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>], <span class="dt">xlab=</span><span class="st">&quot;Ly1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Ln1&quot;</span>)</a>
<a class="sourceLine" id="cb6-2" title="2"><span class="kw">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</a></code></pre></div>
<p><img src="distplus_files/figure-html/lk1mb-2.png" width="700" /></p>
</div>
<div id="redo-clustering-with-alternative-distance-and-agglomeration-method" class="section level3">
<h3>Redo clustering with alternative distance and agglomeration method</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">fit2 =<span class="st"> </span><span class="kw">hclust</span>(cd, <span class="dt">method=</span><span class="st">&quot;ward.D2&quot;</span>)</a>
<a class="sourceLine" id="cb7-2" title="2"><span class="kw">plot</span>(fit2)</a>
<a class="sourceLine" id="cb7-3" title="3"><span class="kw">abline</span>(<span class="dt">h=</span>.<span class="dv">3</span>, <span class="dt">lty=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="distplus_files/figure-html/redo-1.png" width="700" /></p>
</div>
<div id="silhouette-measure" class="section level3">
<h3>Silhouette measure</h3>
<p>From ?silhouette with the cluster library:</p>
<pre><code>For each observation i, the _silhouette width_ s(i) is defined as follows:

     Put a(i) = average dissimilarity between i and all other points of
     the cluster to which i belongs (if i is the _only_ observation in
     its cluster, s(i) := 0 without further calculations).  For all
     _other_ clusters C, put d(i,C) = average dissimilarity of i to all
     observations of C.  The smallest of these d(i,C) is b(i) := \min_C
     d(i,C), and can be seen as the dissimilarity between i and its
     &quot;neighbor&quot; cluster, i.e., the nearest one to which it does _not_
     belong.  Finally,

                   s(i) := ( b(i) - a(i) ) / max( a(i), b(i) ).         
     
     &#39;silhouette.default()&#39; is now based on C code donated by Romain
     Francois (the R version being still available as
     &#39;cluster:::silhouette.default.R&#39;).

     Observations with a large s(i) (almost 1) are very well clustered,
     a small s(i) (around 0) means that the observation lies between
     two clusters, and observations with a negative s(i) are probably
     placed in the wrong cluster.</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">ct1 =<span class="st"> </span><span class="kw">cutree</span>(fit2, <span class="dt">h=</span>.<span class="dv">3</span>)</a>
<a class="sourceLine" id="cb9-2" title="2">c2 =<span class="st"> </span>cnv_log_R</a>
<a class="sourceLine" id="cb9-3" title="3"><span class="kw">rownames</span>(c2) =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">rownames</span>(c2), <span class="kw">as.numeric</span>(ct1))</a>
<a class="sourceLine" id="cb9-4" title="4">sil =<span class="st"> </span><span class="kw">silhouette</span>(ct1, cd)</a>
<a class="sourceLine" id="cb9-5" title="5"><span class="kw">plot</span>(sil)</a></code></pre></div>
<p><img src="distplus_files/figure-html/init-1.png" width="700" /></p>
</div>
<div id="exercises" class="section level3">
<h3>Exercises</h3>
<p><strong>1: install bioDist and vjcitn/csamaDist. Use the code:</strong></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="kw">library</span>(csamaDist)</a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">data</span>(cnv_log_R)</a>
<a class="sourceLine" id="cb10-3" title="3">hc1 =<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(cnv_log_R[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>,]))</a>
<a class="sourceLine" id="cb10-4" title="4">hc2 =<span class="st"> </span><span class="kw">hclust</span>(bioDist<span class="op">::</span><span class="kw">cor.dist</span>(cnv_log_R[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>,]), <span class="dt">method=</span><span class="st">&quot;ward.D2&quot;</span>) </a>
<a class="sourceLine" id="cb10-5" title="5">opar =<span class="st"> </span><span class="kw">par</span>(<span class="dt">no.readonly=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-6" title="6"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">1</span>))</a>
<a class="sourceLine" id="cb10-7" title="7"><span class="kw">plot</span>(hc1, <span class="dt">main=</span><span class="st">&quot;Euc, complete&quot;</span>)</a>
<a class="sourceLine" id="cb10-8" title="8"><span class="kw">plot</span>(hc2, <span class="dt">main=</span><span class="st">&quot;1-Cor, Ward&#39;s D2&quot;</span>)</a>
<a class="sourceLine" id="cb10-9" title="9"><span class="kw">par</span>(opar)</a></code></pre></div>
<p><strong>Comment on the qualitative differences between the clusterings</strong>.</p>
<p><strong>2: Here is how to produce a report on silhouette measurement for the second clustering.</strong></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">assn =<span class="st">  </span><span class="kw">cutree</span>(hc2, <span class="dt">h=</span>.<span class="dv">25</span>)</a>
<a class="sourceLine" id="cb11-2" title="2"><span class="kw">plot</span>(<span class="kw">silhouette</span>( assn, bioDist<span class="op">::</span><span class="kw">cor.dist</span>(cnv_log_R[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>,])))</a></code></pre></div>
<p><strong>Produce a three-cluster partition from <code>hc1</code> and obtain the silhouette display.</strong></p>
</div>
</div>
</div>
<div id="clustering-single-cell-rna-seq" class="section level1">
<h1>Clustering single cell RNA-seq</h1>
<p>This code is taken verbatim from the bluster “diagnostics” vignette.</p>
<div id="acquire-grun-et-als-single-cell-rna-seq-dataset" class="section level2">
<h2>Acquire Grun et al’s single cell RNA-seq dataset</h2>
<p>[Grun 2016] (<a href="https://www.sciencedirect.com/science/article/pii/S1934590916300947" class="uri">https://www.sciencedirect.com/science/article/pii/S1934590916300947</a>) define an algorithm, StemID, that infers candidate multipotent cell populations in the human pancreas.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb12-2" title="2">sce &lt;-<span class="st"> </span><span class="kw">GrunPancreasData</span>()</a></code></pre></div>
<pre><code>## snapshotDate(): 2022-04-26</code></pre>
<pre><code>## see ?scRNAseq and browseVignettes(&#39;scRNAseq&#39;) for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<pre><code>## snapshotDate(): 2022-04-26</code></pre>
<pre><code>## see ?scRNAseq and browseVignettes(&#39;scRNAseq&#39;) for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="co"># Quality control to remove bad cells.</span></a>
<a class="sourceLine" id="cb19-2" title="2"><span class="kw">library</span>(scuttle)</a>
<a class="sourceLine" id="cb19-3" title="3">qcstats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce)</a>
<a class="sourceLine" id="cb19-4" title="4">qcfilter &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(qcstats, <span class="dt">sub.fields=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>)</a></code></pre></div>
<pre><code>## Warning in .get_med_and_mad(metric, batch = batch, subset = subset,
## share.medians = share.medians, : missing values ignored during outlier detection</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">sce &lt;-<span class="st"> </span>sce[,<span class="op">!</span>qcfilter<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb21-2" title="2"></a>
<a class="sourceLine" id="cb21-3" title="3"><span class="co"># Normalization by library size.</span></a>
<a class="sourceLine" id="cb21-4" title="4">sce &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce)</a>
<a class="sourceLine" id="cb21-5" title="5"></a>
<a class="sourceLine" id="cb21-6" title="6"><span class="co"># Feature selection.</span></a>
<a class="sourceLine" id="cb21-7" title="7"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb21-8" title="8">dec &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(sce)</a>
<a class="sourceLine" id="cb21-9" title="9">hvgs &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec, <span class="dt">n=</span><span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb21-10" title="10"></a>
<a class="sourceLine" id="cb21-11" title="11"><span class="co"># Dimensionality reduction.</span></a>
<a class="sourceLine" id="cb21-12" title="12"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb21-13" title="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb21-14" title="14">sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">ncomponents=</span><span class="dv">20</span>, <span class="dt">subset_row=</span>hvgs)</a>
<a class="sourceLine" id="cb21-15" title="15">sce &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce, <span class="dt">subset_row=</span>hvgs)</a></code></pre></div>
</div>
<div id="clustering-using-a-nearest-neighbor-graph-visualization-via-tsne-and-pca" class="section level2">
<h2>Clustering using a nearest-neighbor graph; visualization via TSNE and PCA</h2>
<p>From bluster’s makeSNNGraph help page</p>
<pre><code>   The &#39;makeSNNGraph&#39; function builds a shared nearest-neighbour
   graph using observations as nodes. For each observation, its &#39;k&#39;
   nearest neighbours are identified using the &#39;findKNN&#39; function,
   based on distances between their expression profiles (Euclidean by
   default). An edge is drawn between all pairs of observations that
   share at least one neighbour, weighted by the characteristics of
   the shared nearest neighbors - see &quot;Weighting Schemes&quot; below.

   The aim is to use the SNN graph to perform clustering of
   observations via community detection algorithms in the &#39;igraph&#39;
   package. This is faster and more memory efficient than
   hierarchical clustering for large numbers of observations. In
   particular, it avoids the need to construct a distance matrix for
   all pairs of observations. Only the identities of nearest
   neighbours are required, which can be obtained quickly with
   methods in the &#39;BiocNeighbors&#39; package.</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw">library</span>(bluster)</a>
<a class="sourceLine" id="cb23-2" title="2">mat &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(sce)</a>
<a class="sourceLine" id="cb23-3" title="3">clust.info &lt;-<span class="st"> </span><span class="kw">clusterRows</span>(mat, <span class="kw">NNGraphParam</span>(), <span class="dt">full=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb23-4" title="4">clusters &lt;-<span class="st"> </span>clust.info<span class="op">$</span>clusters</a>
<a class="sourceLine" id="cb23-5" title="5"><span class="kw">table</span>(clusters)</a></code></pre></div>
<pre><code>## clusters
##   1   2   3   4   5   6   7   8   9  10  11  12 
## 285 171 161  59 174  49  70 137  69  65  28  23</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>TSNE, <span class="dt">col=</span>clusters, <span class="dt">pch=</span><span class="dv">19</span>, <span class="dt">main=</span><span class="st">&quot;TSNE&quot;</span>)</a></code></pre></div>
<p><img src="distplus_files/figure-html/graphclust-1.png" width="700" /></p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>PCA, <span class="dt">col=</span>clusters, <span class="dt">pch=</span><span class="dv">19</span>, <span class="dt">main=</span><span class="st">&quot;PCA&quot;</span>)</a></code></pre></div>
<p><img src="distplus_files/figure-html/graphclust-2.png" width="700" /></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw">pairs</span>(<span class="kw">reducedDims</span>(sce)<span class="op">$</span>PCA[,<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>], <span class="dt">col=</span>clusters, <span class="dt">pch=</span><span class="dv">19</span>, <span class="dt">main=</span><span class="st">&quot;PCA&quot;</span>)</a></code></pre></div>
<p><img src="distplus_files/figure-html/graphclust-3.png" width="700" /></p>
</div>
<div id="assessment-via-pairwisemodularity" class="section level2">
<h2>Assessment via pairwiseModularity</h2>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1">g &lt;-<span class="st"> </span>clust.info<span class="op">$</span>objects<span class="op">$</span>graph</a>
<a class="sourceLine" id="cb28-2" title="2">ratio &lt;-<span class="st"> </span><span class="kw">pairwiseModularity</span>(g, clusters, <span class="dt">as.ratio=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb28-3" title="3"></a>
<a class="sourceLine" id="cb28-4" title="4">cluster.gr &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">graph_from_adjacency_matrix</span>(<span class="kw">log2</span>(ratio<span class="op">+</span><span class="dv">1</span>), </a>
<a class="sourceLine" id="cb28-5" title="5">    <span class="dt">mode=</span><span class="st">&quot;upper&quot;</span>, <span class="dt">weighted=</span><span class="ot">TRUE</span>, <span class="dt">diag=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb28-6" title="6"></a>
<a class="sourceLine" id="cb28-7" title="7"><span class="co"># Increasing the weight to increase the visibility of the lines.</span></a>
<a class="sourceLine" id="cb28-8" title="8"><span class="kw">set.seed</span>(<span class="dv">1100101</span>)</a>
<a class="sourceLine" id="cb28-9" title="9"><span class="kw">plot</span>(cluster.gr, <span class="dt">edge.width=</span>igraph<span class="op">::</span><span class="kw">E</span>(cluster.gr)<span class="op">$</span>weight<span class="op">*</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb28-10" title="10">    <span class="dt">layout=</span>igraph<span class="op">::</span>layout_with_lgl)</a></code></pre></div>
<p><img src="distplus_files/figure-html/dopairw-1.png" width="700" /></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer></div>

  

  

  </body></html>
